src/mds/MDSUDPEndpoint.cc

```
@@ -0,0 +1,91 @@
+  }
+  std::shared_ptr<MDSUDPEndpoint> new_endpoint =
+      MDSUDPEndpoint::create(cct, name, ip, port);
+  if (!new_endpoint) {
+    return;
+  }
+  endpoints[name] = new_endpoint;
+}
+
+void MDSUDPManager::remove_endpoint(const std::string &name) {
+  std::unique_lock<std::shared_mutex> lock(endpoint_mutex);
+  auto it = endpoints.find(name);
+  if (it != endpoints.end()) {
+    endpoints.erase(it);
+  }
+}
+
+MDSUDPEndpoint::MDSUDPEndpoint(const std::string &name, const std::string &ip,
+                               int port)
+    : name(name), socket(io_context), ip(ip), port(port),
+      endpoint(boost::asio::ip::address::from_string(ip), port) {
+  try {
+    boost::system::error_code ec;
+    socket.open(boost::asio::ip::udp::v4(), ec);
+    if (ec) {
+      throw std::runtime_error(ec.message());
+    }
+  } catch (const std::exception &e) {
+    ldout(cct, 0) << "Error occurred while opening UDP socket with error:"
+                  << e.what() << dendl;
+    throw;
+  }
+}
+
+std::shared_ptr<MDSUDPEndpoint> MDSUDPEndpoint::create(CephContext *_cct,
+                                                       const std::string &name,
+                                                       const std::string &ip,
+                                                       int port) {
+  if (_cct && !MDSUDPEndpoint::cct) {
+    MDSUDPEndpoint::cct = _cct;
+  }
+  try {
+    std::shared_ptr<MDSUDPEndpoint> endpoint =
+        std::make_shared<MDSUDPEndpoint>(name, ip, port);
+    return endpoint;
+  } catch (...) {
+  }
+  return nullptr;
+}
+
+uint64_t MDSUDPEndpoint::publish_internal(
+    const std::shared_ptr<MDSNotificationMessage> &message) {
+  boost::system::error_code ec;
+  socket.send_to(boost::asio::buffer(message->message, message->message_length),
+                 endpoint, 0, ec);
+  if (ec) {
+    ldout(cct, 0) << "Error occurred while sending notification having seq_id="
+                  << message->seq_id << ":" << ec.message() << dendl;
```

IMO better to return the error to a  caller